### ISSUE TITLE
dvc.yaml ignored if entire dvclive folder is tracked as a stage output

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvclive#456]
Reproduction script below. `exp show` shows nothing in the workspace once `dvclive` is added as a stage output.

```bash
$ git clone git@github.com:iterative/dvclive-exp-tracking.git

$ cd dvclive-exp-tracking

$ python train.py

$ dvc exp show
 ────────────────────────────────────────────────────────────────────────────────────
  Experiment                 Created        train_loss   epoch   step   encoder_size
 ────────────────────────────────────────────────────────────────────────────────────
  workspace                  -                0.026054       4    499   64
  main                       Feb 01, 2023            -       -      -   -
  └── 4f01b48 [leafy-wage]   01:10 PM         0.026054       4    499   64
 ────────────────────────────────────────────────────────────────────────────────────

$ dvc stage add -n train -d train.py -O dvclive python train.py

$ dvc exp show
 ──────────────────────────────────────────────────────────────────────────────────────>
  Experiment                 Created        train_loss   epoch   step   encoder_size   >
 ──────────────────────────────────────────────────────────────────────────────────────>
  workspace                  -                       -       -      -   -              >
  main                       Feb 01, 2023            -       -      -   -              >
  └── 4f01b48 [leafy-wage]   01:10 PM         0.026054       4    499   64             >
 ──────────────────────────────────────────────────────────────────────────────────────>
```

This happens because `dvc.yaml` is within the output, which DVC doesn't seem to like. You can avoid it by writing the stage like:

```bash
$ dvc stage add -f -n train -d train.py -O dvclive/metrics.json -o dvclive/plots python train.py
```

However, it's probably not safe to rely on users doing this. Some ways we can remedy it:
1. Write to root `dvc.yaml` (see discussion in #381).
2. Fix on DVC side to look for `dvc.yaml` inside an output (not sure this makes sense since it could include pipeline stages, which would be odd to have inside an output).
3. Other ideas?

**Comment by daavoo** [comment]
> Other ideas?

Don't mention `dvclive` folder at all in the root `dvc.yaml`? What would be the downside?

**Comment by dberenbaum** [comment]
That's true. It's not likely to be a dependency of another stage. Regardless, not sure if documentation is enough here or if it's worth taking some action to prevent possible confusion for users who try it.

**Comment by shcheklein** [comment]
A agree with @daavoo , I also don't see a good reason to mention it at all, tbh. Except if metrics become really large, but probably solving that with a granular `-O dvclive/metrics.json -o dvclive/plots` should be fine?

**Comment by dberenbaum** [comment]
So default should be to suggest `$ dvc stage add -f -n train -d train.py` here?

**Comment by shcheklein** [comment]
Yep, sounds like that. That's what we do in VS Code btw with @sroy3, at least for now.

**Comment by dberenbaum** [comment]
Thinking of the onboarding experience, maybe it's not bad to default to including `-O dvclive/metrics.json -o dvclive/plots` to introduce the concept of outputs and a more typical pipeline stage structure. Having no outputs could raise more questions. WDYT?

**Comment by dberenbaum** [comment]
In https://github.com/iterative/dvc.org/pull/4389#discussion_r1142591545, tried to put `dvclive/dvc.yaml` into `.dvcignore` to fix this issue, but then `dvc exp show` doesn't read the ignored `dvclive/dvc.yaml` and ignores all the DVCLive metrics.

Should DVC either read `dvc.yaml` files even within outputs, or else can it read `dvc.yaml` files even if they are inside ignored paths?

**Comment by dberenbaum** [comment]
Closing since we have aligned on leaving this out of `dvc.yaml`.

**Comment by dberenbaum** [comment]
A user ran into this. They don't want to store all dvclive changes in git because it looks like a ton of file changes that aren't of interest, but if they don't git track at least the `dvc.yaml` file, dvclive doesn't work as expected.

The best thing might be to force add the dvclive `dvc.yaml` to git no matter what.

**Comment by dberenbaum** [comment]
I also think that a better solution for https://github.com/iterative/dvclive/issues/568 might be to simply have a `cache=True/False` option for the entire DVCLive folder, but that requires resolving this issue.

**Comment by dberenbaum** [comment]
@skshetry Maybe you have ideas here. Is it expected that dvc will/will not collect from a dvc.yaml file inside a dvc-tracked output?

**Comment by dberenbaum** [comment]
Makes sense. I guess we can deprioritize this after https://github.com/iterative/dvc.org/pull/4674 since it at least explains the issue and how to work around it.

**Comment by skshetry** [comment]
I think the issue is dvclive writing to a `dvc.yaml` in a separate directory. It should follow the conventions of `dvc`.

**Comment by dberenbaum** [comment]
@skshetry If you define the stage in the root `dvc.yaml`, and running the stage modifies that same `dvc.yaml`, do you think that follows the conventions of `dvc`? It may also be problematic to edit that file while running the stage, and there are also known issues with trying to modify the user's `dvc.yaml` (for example, https://github.com/iterative/dvc/issues/5736, https://github.com/iterative/dvc/issues/8720).

**Comment by skshetry** [comment]
Modifying `dvc.yaml`, whether it's in the root or `dvclive` directory, is problematic as `dvc.yaml` file is supposed to be static.

`dvc` rarely modifies `dvc.yaml` itself (and, only in select commands like `stage add`/`freeze`/`unfreeze`).

Unless the modification to `dvc.yaml` is of invalid format, it should be okay to modify. `repro` would not complain or notice.

There are two things to keep in mind however:
1. dvc does allow multiple `repro` to run in parallel, so some modifications might get noticed by other repro commands.
2. `post_live_metrics` might read the latest changes, which is intended.

These problems happen regardless of where the modifications are being made.

We do have to remember that while modifying `dvc.yaml`, it should be done atomically as possible, which we are not doing today.

**Comment by dberenbaum** [comment]
> We do have to remember that while modifying `dvc.yaml`, it should be done atomically as possible, which we are not doing today.

Why? So you think DVCLive should be repeatedly opening and writing to `dvc.yaml`?

**Comment by skshetry** [comment]
> Why? So you think DVCLive should be repeatedly opening and writing to dvc.yaml?

That was a general comment unrelated to dvclive. `dvc` does not save atomically either, but we should. Otherwise, another dvc process might try to read `dvc.yaml` while the process in the middle of the write (which might be broken).

`dvc` has to open and then write repeatedly to preserve formatting anyway. I see `dvclive` does not do a round-trip, but it should if it needs to preserve formatting. I can't answer about "repeatedly" though, depends on what dvclive is trying to do.

**Comment by skshetry** [comment]
Okay, now I understand your comment. By "atomically", I mean that the partial writes should not be visible to other processes. So, we should be writing to a temporary file somewhere and then replacing/renaming that file atomically.

**Comment by daavoo** [comment]
Currently, `dvclive` doesn't update. It completely overwrites the entire file in a single call

**Comment by skshetry** [comment]
yes, what I said above applies to updates or overwrites or even a new file.

What I mean is:
```python
with open("dvc.yaml", mode="w") as f:
   pass
```

As soon as it enters the block, it'll truncate and create an empty file which might be visible to the other processes. Or, multiple `write` might be required to completely write to that file.

Again, it's a general comment regarding safety. We allow other commands to run in parallel.

**Comment by dberenbaum** [comment]
These are all the reasons we have it as a separate file now, though. We don't have to worry about round-trips or atomicity.

Not that I don't see benefits to having everything in the root `dvc.yaml`, but I'm not sure it's worthwhile for all the headaches I anticipate it will cause.

**Comment by skshetry** [comment]
> These are all the reasons we have it as a separate file now, though. We don't have to worry about round-trips or atomicity.

DVC will read them anyway.

**Comment by dberenbaum** [comment]
Yeah, but all these concerns are about writing.

**Comment by dberenbaum** [comment]
Another concern with leaving the dvclive folder untracked is what happens when reproducing an experiment. If I run `dvc repro/exp run` and dvc is able to recover a previous run from the cache, it will fail to update the dvclive outputs to match that cached run. I would also want to recover the cached `dvclive/dvc.yaml` since that has the right config for that run, so moving everything to the base `dvc.yaml` doesn't fully "fix" this issue IMO.

**Comment by dflatow** [comment]
OK so I maybe running into something related to this (apologies if user error). I'm using pipelines. Basically I **thought** I should write plots and metrics to a folder specified as an output of a stage. However, DVC doesn't like this and plots don't show up because the `dvc.yaml` is in a folder that's an output. After scratching my head for an unreasonably long time I finally saw this issue and realized that having the output folder as the same as where dvclive writes is no bueno. 

FWIW I found this really unintuitive... 

Is it possible to warn the user (via console warning in dvclive) that they shouldn't use `dir=somedir` if `somedir` is listed as an output in any pipeline stage?

**Comment by dflatow** [comment]
Also there seems to be some strange behavior (in terms of plots not showing up in experiments) if some directory was previously listed as an output (but no longer)

**Comment by dberenbaum** [comment]
> Also there seems to be some strange behavior (in terms of plots not showing up in experiments) if some directory was previously listed as an output (but no longer)

@dflatow If you are able to narrow that down and open a separate issue, it would be much appreciated!

**Comment by daavoo** [comment]
> Is it possible to warn the user (via console warning in dvclive) that they shouldn't use dir=somedir if somedir is listed as an output in any pipeline stage?

This makes sense to me, wdyt @dberenbaum ?

**Comment by mnrozhkov** [comment]
Hey guys! It's a long discussion already, but let me add my 2c :)  I am also having confusion about generated `dvc.yaml`. 
Let me point out a few perspectives

1. `dvc.yaml` for me is a config for DVC pipelines! When I see such a file, I expected to find a pipeline configuration inside. 
2. The `dvc.yaml` file generated by DVCLive looks like a `dvc.lock` file generated by DVC. It contains metadata to be used by  DVC tools, not intended to be changed by a user. All artifacts configurations are stored and updated in the code! 

Is it possible that all these confusions are due to wrong naming? And, we should not have a second `dvc.yaml` at all? 
Could we have a separate `.lock` (aka `artifacts.lock`) to keep all artifacts metadata, versioned by Git? 

Here are some thoughts on a unified workflow for DVC & DVCLive:  

- Let's imagine we store all metadata about artifacts in `.dvc/artifacts.lock`
- Artifacts specified in dvc.yaml pipeline → write meta to `.dvc/artifacts.lock`
- Artifacts saved via DVCLive → write meta to `.dvc/artifacts.lock`
- `cache=True/False` in DVCLive works the same as at `dvc.yaml` (all files/folders can be tracked by DVC)
- All artifacts can be specified as `dvc.yaml:outs/deps` in → DVC reads from `.dvc/artifacts.lock`
- Experiments & Plots UI → read from `.dvc/artifacts.lock`

Summary on `artifacts.lock`

- always tracked by Git
- unifies all artifacts (metrics, plots, images, models…) generated
- have a “stage” field to link an artifact to a stage generated by the artifact

Example for a `dvc.yaml` 
```yaml
stages:
  train:
    cmd: python ...
    metrics:
    - metrics.json                   # OR reference by name 'metrics:train_metrics'

  validate:
    cmd: yolo ...
    deps:
    - metrics.json                  # OR reference by name of the metrics metrics:train_metrics' 
```

```yaml
metrics:
  train_metrics:          # name
     stage: train         # stage
     path: metrics.json   # path 
     hash: md5
     md5: 5725466dfe...
     ...                  # other fields
plots:
  mae:
    type: plots/metrics
    hash: md5
    md5: 5725466dfe...
    ...

artifacts:
  pool-segmentation:
    stage: validate
    path: artifacts/best.pt
    type: model
    hash: md5
    md5: 5725466dfe...
    ...
```

**Comment by dberenbaum** [comment]
> 1. `dvc.yaml` for me is a config for DVC pipelines! When I see such a file, I expected to find a pipeline configuration inside.

Makes sense since that's what it was for a long time. Now it does [more](https://dvc.org/doc/user-guide/project-structure/dvcyaml-files), which is needed outside dvclive (for example, to configure custom plots or add custom model artifacts), so we need to educate that it's more than a pipeline file.

> 2\. The `dvc.yaml` file generated by DVCLive looks like a `dvc.lock` file generated by DVC. It contains metadata to be used by  DVC tools, not intended to be changed by a user. All artifacts configurations are stored and updated in the code!

Right, it's auto-generated and not intended to be changed by a user, but it contains the same info as any other `dvc.yaml` file (specifying plots fields, artifact names/types/labels, etc.) and doesn't have quite the same kind of info as a file like `dvc.lock` (it doesn't contain hash, md5, etc.).

> Is it possible that all these confusions are due to wrong naming? And, we should not have a second `dvc.yaml` at all?
> Could we have a separate `.lock` (aka `artifacts.lock`) to keep all artifacts metadata, versioned by Git?

> * BONUS: no need to generate a separate `.dvc` file for each artifact (`artifacts.lock` contains hashes)

If I understand your proposal, we may not need a new file, since this is basically what `dvc.lock` does. We have had discussions about moving away from `.dvc` files and consolidating all this info into `dvc.lock`. This whole discussion is closely related to #648, and one of the suggestions there is to keep all info in one lock file and get rid of those `.dvc` files.

Note that we would still need a `dvc.yaml` file to specify info like what field to use as the x-axis for the plot. This is what `dvc.yaml` does today, and I don't yet see a good reason to that info outside `dvc.yaml` (users still need to be able to configure those fields, both within stages and for standalone plots, artifacts. etc.).

**Comment by mnrozhkov** [comment]
Thanks @dberenbaum! I just ask the team to consider the user experience they had with `dvc.yaml` for several years! 

If we move all metadata, including what field to use as the x-axis for the plot, into some `.lock` file, it would be enough. 

Then users may choose the way to specify the
1. `dvc.yaml` with a pipeline
2. Python API - DVCLive 

This seems could solve a lot of uncertainties and simplify workflow.

**Comment by dflatow** [comment]
Just some supporting evidence for this:

> Basically I thought I should write plots and metrics to a folder specified as an output of a stage.

Found this in the documentation:
https://dvc.org/doc/user-guide/pipelines/defining-pipelines
![image](https://github.com/iterative/dvclive/assets/5514147/a51294a2-8c93-4347-a6d1-0079b806ce1c)

Seems to suggest that you should in fact make plots in directories that are listed as output directories (which DVC seems not to like):

**Comment by dberenbaum** [comment]
> Seems to suggest that you should in fact make plots and metrics in directories that are listed as output directories (which DVC seems not to like -- especially metrics).

@dflatow Would you mind providing a specific example, preferably in a separate issue (show the error you get, the `dvc.yaml`, any other relevant details)? You should be able to make plots and metrics in output directories.

**Comment by daavoo** [comment]
We now (main branch, waiting for 3.0 release) write to root `dvc.yaml` by default
